server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightPhase.java

```diff
@@ -28,34 +29,83 @@
 import org.elasticsearch.index.query.QueryShardContext;
 import org.elasticsearch.search.SearchShardTarget;
 import org.elasticsearch.search.fetch.FetchSubPhase;
+import org.elasticsearch.search.fetch.FetchSubPhaseProcessor;
 import org.elasticsearch.search.internal.SearchContext;
 
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.LinkedHashMap;
 import java.util.Map;
+import java.util.function.Function;
 
 public class HighlightPhase implements FetchSubPhase {
+
     private final Map<String, Highlighter> highlighters;
 
     public HighlightPhase(Map<String, Highlighter> highlighters) {
         this.highlighters = highlighters;
     }
 
     @Override
-    public void hitExecute(SearchContext context, HitContext hitContext) {
+    public FetchSubPhaseProcessor getProcessor(SearchContext context) {
         if (context.highlight() == null) {
-            return;
+            return null;
         }
-        hitExecute(context.shardTarget(), context.getQueryShardContext(), context.parsedQuery().query(), context.highlight(), hitContext);
+
+        return getProcessor(context.getQueryShardContext(), context.shardTarget(), context.highlight(), context.parsedQuery().query());
+    }
+
+    public FetchSubPhaseProcessor getProcessor(QueryShardContext qsc, SearchShardTarget target, SearchHighlightContext hc, Query query) {
+        Map<String, Function<HitContext, FieldHighlightContext>> contextBuilders = contextBuilders(qsc, target, hc, query);
+        return new FetchSubPhaseProcessor() {
+            @Override
+            public void setNextReader(LeafReaderContext readerContext) {
+
+            }
+
+            @Override
+            public void process(HitContext hitContext) {
+                Map<String, HighlightField> highlightFields = new HashMap<>();
+                for (String field : contextBuilders.keySet()) {
+                    FieldHighlightContext fieldContext = contextBuilders.get(field).apply(hitContext);
+                    Highlighter highlighter = getHighlighter(fieldContext.field);
+                    if ((highlighter.canHighlight(fieldContext.fieldType) == false) && fieldContext.fromWildcard) {
```

Comment:
Could we perform this check only once for the whole subphase (maybe by reworking the `contextBuilders` method)? That would let us avoid adding the `fieldNameContainsWildcards` field to `FieldHighlightContext`.